### PR TITLE
Fix null-bounds calculation for ranged window queries

### PR DESCRIPTION
**Pull request by mythrocks**
This commit fixes an off-by-one error in `grouped_time_range_rolling_window()`, for cases where the order-by column (i.e. the timestamp) contains null values.

The error was reported when running Spark with the `spark-rapids` plugin enabled, manifesting in a process crash:
```
21/03/11 04:26:51 ERROR TaskSetManager: Task 3 in stage 4.0 failed 1 times; aborting job
org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 4.0 failed 1 times, most recent failure: Los
t task 3.0 in stage 4.0 (TID 5, 10.0.0.23, executor driver): ai.rapids.cudf.CudfException: for_each: failed to synchronize: cud
aErrorLaunchFailure: unspecified launch failure
        at ai.rapids.cudf.Table.timeRangeRollingWindowAggregate(Native Method)
        at ai.rapids.cudf.Table.access$3100(Table.java:45)
        at ai.rapids.cudf.Table$AggregateOperation.aggregateWindowsOverTimeRanges(Table.java:2258)
        at com.nvidia.spark.rapids.GpuWindowExpression.$anonfun$evaluateRangeBasedWindowExpression$2(GpuWindowExpression.scala:
233)
        at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
        at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
        at com.nvidia.spark.rapids.GpuWindowExpression.withResource(GpuWindowExpression.scala:132)
        at com.nvidia.spark.rapids.GpuWindowExpression.$anonfun$evaluateRangeBasedWindowExpression$1(GpuWindowExpression.scala:
224)
```

It is interesting that a fair number of the 870+ tests from 75 test cases in `grouped_rolling_test.cpp` exercise this code path. `cuda-memcheck` still does not seem to detect the corruption in `CountMultiGroupTimestampASCNullsFirst`, for instance:
```
✗ cuda-memcheck gtests/GROUPED_ROLLING_TEST --gtest_filter=TypedNullTimestampTestForRangeQ eries/0.CountMultiGroupTimestampASCNullsFirst
========= CUDA-MEMCHECK
Note: Google Test filter = TypedNullTimestampTestForRangeQueries/0.CountMultiGroupTimestampASCNullsFirst
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TypedNullTimestampTestForRangeQueries/0, where TypeParam = signed char
[ RUN      ] TypedNullTimestampTestForRangeQueries/0.CountMultiGroupTimestampASCNullsFirst
[       OK ] TypedNullTimestampTestForRangeQueries/0.CountMultiGroupTimestampASCNullsFirst (663 ms)
[----------] 1 test from TypedNullTimestampTestForRangeQueries/0 (663 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (663 ms total)
[  PASSED  ] 1 test.
========= ERROR SUMMARY: 0 errors
```

### CODE REVIEW

**Comment by mythrocks** [comment]
I've raised this against `branch-0.18` since this is a bug in a released version of `libcudf`. The user will experience application crashes or data corruption. I'm not sure of the best way to update a released CUDF version.

I'd be happy to retarget this PR to `branch-0.19` if that's preferable.

**Comment by kkraus14** [comment]
See the hotfix process here: https://docs.rapids.ai/releases/hotfix/

cc @jrhemstad @harrism for visibility and to weigh in on if this is hotfix worthy

**Comment by kkraus14** [comment]
In addition to the fix we should add a test that captures this behavior.

**Comment by mythrocks** [comment]
Hello, @kkraus14. Thank you for your advice.
> In addition to the fix we should add a test that captures this behavior

This is an off-by-1 error that should have been triggered by tests that currently exist in `grouped_rolling_test.cpp`. (`CountMultiGroupTimestampASCNullsFirst`, for instance.) I am as yet unable to reliably reproduce the crash from unit-tests, short of using a manually extracted Parquet file. The crash isn't a consistent one, outside of running `cuda-memcheck` as part of the test.

Please pardon my temerity, but I'm not sure I would have written a test for this case apart from the ones we already have. :/

**Comment by mythrocks** [comment]
> See the hotfix process here: https://docs.rapids.ai/releases/hotfix/

The process prescribes the following:
> 2. Create your branch from the `branch-M.B` branch

`M.B` is the next minor release i.e. `0.19`. I will re-target this PR for `branch-0.19`, and raise a separate one for `main`.

**Comment by mythrocks** [comment]
The PR has been re-targeted to `branch-0.19`. I'll wait on the team's evaluation of whether this needs a hotfix. IMHO, it probably does, since this is a memory corruption.

File: cpp/src/rolling/grouped_rolling.cu
Comment (jrhemstad) [inline cpp/src/rolling/grouped_rolling.cu:387]: This is surprising. Without digging into the details, I presume this is a common pattern of a vector of offsets such that the size of any group should be determined by `group_offsets[i+i] - group_offsets[i]`? In which case I'd actually expect the number of groups to be `group_offsets.size() - 1`.

**Comment by jrhemstad** [comment]
The existence of memory corruption bugs has usually not warranted a hotfix in the past unless it was an extremely pervasive bug, e.g. something like a function in `column_view` caused a segfault. This seems contained to a single feature, and I don't know how commonly used that feature is.

**Comment by revans2** [comment]
> The existence of memory corruption bugs has usually not warranted a hotfix in the past unless it was an extremely pervasive bug, e.g. something like a function in `column_view` caused a segfault. This seems contained to a single feature, and I don't know how commonly used that feature is.

The issue started to show up in the Spark integration tests nearly 100% of the time they run, after we changed our hash partitioning to match what Spark does.  Because the crash depends on very specific data conditions and also reading uninitialized memory, the conditions to make it happen are not easy to reproduce, unless we run with cuda-memcheck with RMM pooling disabled. That changes a potential data corruption bug into a guaranteed crash.

Silent data corruption, is a stop the world kind of bug for the Spark plugin. Our customers have to trust that we will give them the correct answer. If we do not re-spin a 0.18, the only real choice left to us is to disable all range based window operations if the order by column is nullable. Because Spark is not very good at figuring out ahead of time if something really is nullable or not it would effectively mean that we cannot support any range based window queries at all, which would impact almost all of our customers.

File: cpp/src/rolling/grouped_rolling.cu
Comment (revans2) [reply cpp/src/rolling/grouped_rolling.cu:387]: I don't think `num_groups` is an accurate name any more. If it is holding the number of groups then it needs some comments explaining why that is true. If it is not then it needs to be renamed.

Line 399 is also very concerning in relation to this change, and I think is what @jrhemstad is pointing out in his comment.

```
auto group_end             = d_group_offsets[group_label + 1];
```

How are we not walking off the end of `d_group_offsets`? `group_label` is the current index between [0, group_offsets.size() + 1), which means the largest `group_label` value will be `group_offsets.size()` which should already be larger than anything `d_group_offsets` could hold, and now we are going one more to it.  Doesn't this mean that before we were reading data off the end of d_group_offsets by 1 value, and now after this change we are reading it up to 2 values past the end?

**Comment by raydouglass** [comment]
> The PR has been re-targeted to `branch-0.19`.

Sorry about this, but the hotfix process was documented incorrectly. It's been updated now.
Here's the change: https://github.com/rapidsai/docs/pull/172

In short, this PR should target `branch-0.18`.

**Comment by revans2** [comment]
After talking with management I think the solution for spark that we are likely to go with is to configure it off by default and let customers opt in.  Still need to check with more people in the management chain, but that looks like it is the likely path forward.

**Comment by kkraus14** [comment]
@revans2 I'm kicking off conversations with @rapidsai/ops and others to scope the level of effort a hotfix would entail. I think your justification for why you'd want this to be a hotfix makes sense as Window functions are much more common in the Spark and SQL worlds than the Python world.

If you can hold off on making a directional decision for Spark-RAPIDS for a bit, please do.

File: cpp/src/rolling/grouped_rolling.cu
Comment (mythrocks) [reply cpp/src/rolling/grouped_rolling.cu:387]: > In which case I'd actually expect the number of groups to be group_offsets.size() - 1.

@jrhemstad, you're right. It should have been `group_offsets.size()-1`. I'm not sure how I managed to flip it. (I think I muffed that up when redoing it for branch-0.19. Fatigue setting in. :|)

It also looks like I should be re-re-targeting for `branch-0.18`, as per @raydouglass's advice. 

Permission to force-push after retesting?

**Comment by raydouglass** [comment]
This PR needs to be rebased onto `branch-0.18` to eliminate the extra commits from `branch-0.19`.

File: cpp/src/rolling/grouped_rolling.cu
Comment (mythrocks) [reply cpp/src/rolling/grouped_rolling.cu:387]: > I don't think num_groups is an accurate name any more.

`num_groups` is what it is. I had this right in the `0.18` version of the fix, and fumbled when redoing it for `branch-0.19`.

**Comment by mike-wendt** [comment]
> This PR needs to be rebased onto `branch-0.18` to eliminate the extra commits from `branch-0.19`.

Can also checkout `branch-0.18` and cherry-pick your change if the rebase gives you any trouble @mythrocks

**Comment by mythrocks** [comment]
> Can also checkout branch-0.18 and cherry-pick your change if the rebase gives you any trouble @mythrocks

Hey, @raydouglass. I was planning on doing just that, instead of a rebase. I need a moment to retest, `cuda-memcheck`, etc.
I'm on it.

File: cpp/src/rolling/grouped_rolling.cu
Comment (mythrocks) [reply cpp/src/rolling/grouped_rolling.cu:387]: > mythrocks force-pushed the mythrocks:fix-null-bounds-calculation branch from f4a0bdf to 7ca66d5 9 hours ago

Commit f4a0bdf had this right.

**Comment by mythrocks** [comment]
> I'm on it.

I have pushed the right fix. (I had my old branch around. I needed to retest through Spark, `cuda-memcheck`, etc.)

**Review by revans2 — APPROVED** [review]
This change fixes the issue. I have run the tests that showed the problem several times and could not reproduce it.

**Review by jrhemstad — APPROVED** [review]

**Comment by kkraus14** [comment]
Will let @harrism take a look before we move forward in starting the hotfix release process. Thanks all.

**Comment by mike-wendt** [comment]
Working on getting the tests running on this as we don't have testing enabled for this branch as it is not active

**Comment by raydouglass** [comment]
rerun tests

**Comment by kkraus14** [comment]
Python build failures are expected because of the Dask branch changes. Do we need to resolve these things on 0.18?

**Comment by mike-wendt** [comment]
> Python build failures are expected because of the Dask branch changes. Do we need to resolve these things on 0.18?

@mythrocks can you patch the contents of #7535 onto your PR? This way we can get the Python builds and GPU tests to work now that Dask has changed their default branch?

**Comment by kkraus14** [comment]
Needs changes from https://github.com/rapidsai/cudf/pull/7532 as well. I'll add them.

**Comment by kkraus14** [comment]
@mike-wendt changes should be added

**Review by vuule — APPROVED** [review]

**Comment by mike-wendt** [comment]
Will merge when @raydouglass returns this afternoon

**Comment by mythrocks** [comment]
> > Python build failures are expected because of the Dask branch changes. Do we need to resolve these things on 0.18?
> 
> @mythrocks can you patch the contents of #7535 onto your PR? This way we can get the Python builds and GPU tests to work now that Dask has changed their default branch?

@mike-wendt: Acknowledging here with gratitude that @kkraus14 updated the PR before I read this message.

**Comment by kkraus14** [comment]
```
13:15:12 ImportError: cannot import name 'stringify_path' from 'dask.bytes.core' (/opt/conda/envs/rapids/conda-bld/dask-cudf_1615572592450/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/site-packages/dask/bytes/core.py)
```
Needs the fix from #7580

**Comment by harrism** [comment]
I'm concerned that our guests failed to catch an error like this. @mythrocks if this code path is exercised by multiple tests as you say, why do none of them fail? This PR should add a test that fails without the fix and succeeds after the fix.

**Comment by kkraus14** [comment]
> I'm concerned that our guests failed to catch an error like this. @mythrocks if this code path is exercised by multiple tests as you say, why do none of them fail? This PR should add a test that fails without the fix and succeeds after the fix.

Lets add the test in 0.19 so as to not delay the hotfix any further?

**Comment by mythrocks** [comment]
>> This PR should add a test that fails without the fix and succeeds after the fix.

It hasn't been from want of effort. Outside of exercising this code path from Spark through a handcrafted Parquet file, I have been unable to reproduce the crash, even by constructing the same input columns programmatically.

> Lets add the test in 0.19 so as to not delay the hotfix any further?

I have filed #7590 to add a test to `branch-0.19`. I will try hit this ASAP.

**Pull request merged**
